pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -59,56 +76,96 @@ func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenI
 	return p.svc.GetAuthorizationToken(input)
 }
 
+// lazyEcrProvider is a DockerConfigProvider that creates on demand an
+// ecrProvider for a given region and then proxies requests to it.
+type lazyEcrProvider struct {
+	region         string
+	regionURL      string
+	actualProvider *credentialprovider.CachingDockerConfigProvider
+}
+
 // ecrProvider is a DockerConfigProvider that gets and refreshes 12-hour tokens
 // from AWS to access ECR.
 type ecrProvider struct {
-	getter tokenGetter
+	region    string
+	regionURL string
+	getter    tokenGetter
 }
 
+// Init creates a lazy provider for each AWS region, in order to support
+// cross-region ECR access. They have to be lazy because it's unlikely, but not
+// impossible, that we'll use more than one.
 // Not using the package init() function: this module should be initialized only
 // if using the AWS cloud provider. This way, we avoid timeouts waiting for a
 // non-existent provider.
 func Init() {
-	credentialprovider.RegisterCredentialProvider("aws-ecr-key",
-		&credentialprovider.CachingDockerConfigProvider{
-			Provider: &ecrProvider{},
-			// Refresh credentials a little earlier before they expire
+	for _, region := range AWSRegions {
+		credentialprovider.RegisterCredentialProvider("aws-ecr-"+region,
+			&credentialprovider.CachingDockerConfigProvider{
+				Provider: &lazyEcrProvider{
+					region:    region,
+					regionURL: fmt.Sprintf(registryURLTemplate, region),
+				},
+				// This is going to be just a lazy proxy to the real ecrProvider.
+				// It holds no real credentials, so refresh practically never.
+				Lifetime: 365 * 24 * time.Hour,
```

Comment:
Didn't checked the code in credentialprovider, but I *guess* it is not an optional argument. And is registering this lazy provider as the provider. If one is actually used, then the real provider is created.

But maybe I'm not understanding correctly. But this makes me wonder, why not do the lazyness on credentialprovider? Or we don't want the lazyness at that abstraction level? (never even looked at the code, so it might be a totally stupid question, sorry in advance)